lib/doorkeeper/server.rb

```diff
@@ -37,8 +42,7 @@ def resource_owner
     end
 
     def credentials
```

Comment:
This is used by the `client` method above, however, we probably need to change that method to:

```
def client
  @client ||= OAuth::Client.find(credentials.uid)
end
```
and have:
```
def client_authenticated?
  return unless client
  return client if credentials.blank? && !client.confidential?
  return client if client.secret_matches?(credentials.secret)
end
```

i.e., we're separating the "getting of the client" from the "authenticating of a client", such that for things like token revocation, we can just get the client, authenticate it if necessary, but then assert that the access token or refresh token was issued to that client.

I'm not 100% sure on the exact specifics here though.

Reply:
Now that I think about it, maybe as it is is better? You only get a `client` when it's authenticated, even if there were credentials on the request. Maybe it's important to know _how_ the client is authenticated, I'm not sure?